### DISCUSSION TITLE
Looking for a way to export Jacobian from pygimi to pytorch

### DISCUSSION THREAD

**Discussion by laughingrice** [gimli-org/pyGIMLi#277] [Q&A]
I'm trying to use pygimli as a piece in a neural network set up with pytorch
After some work, I figured out how to generate the travel time tomography Jacobian (of type RSparseMapMatrix)

```
tt = TravelTimeManager()
tt.setData(scheme)
tt.setMesh(mesh)
tt.fop.createJacobian(1/vel)

J = tt.fop.jacobian()
```

Now I'm trying to figure out how to get it out of pygimli and into PyTorch. I don't see access into the row idx, col idx, and val buffers, and it looks like the way to go is via

```
J.fillArrays
```

but I cannot figure out how to build a buffer that I can use (ideally pass the tensor buffer from PyTorch or a NumPy array into pygimli).

I tried either sending a NumPy array directly or IVector and RVector but getting an error that the types do not match. As the function is expecting a class, it does not seem that going down the ctypes route is the way to go either.

```
r = pg.matrix.IVector(J.nVals())
c = pg.matrix.IVector(J.nVals())
v = pg.matrix.RVector(J.nVals())

J.fillArrays(v,r,c)
```

output is

```
---------------------------------------------------------------------------
ArgumentError                             Traceback (most recent call last)
<ipython-input-57-1138b03bb0c7> in <module>
----> 1 J.fillArrays(v,r,c)

ArgumentError: Python argument types in
    RSparseMapMatrix.fillArrays(RSparseMapMatrix, RVector, IVector, IVector)
did not match C++ signature:
    fillArrays(GIMLI::SparseMapMatrix<double, unsigned long> {lvalue}, GIMLI::Vector<double> {lvalue} vals, GIMLI::Vector<unsigned long> {lvalue} rows, GIMLI::Vector<unsigned long> {lvalue} cols)
```

**Comment by laughingrice** [comment]
Figured it out

```
coo = pg.utils.sparseMatrix2coo(J)
ind = torch.LongTensor(np.stack((coo.row, coo.col)))
val = torch.FloatTensor(coo.data)
tJ = torch.sparse.FloatTensor(ind, val, torch.Size(coo.shape))
```

**Comment by carsten-forty2** [comment]
good, thats the supposed way to convert core matrices to numpy/scipy matrices.

Interesting project btw.
